### PR TITLE
Feat/#323 precompiles ec add

### LINKED ISSUE #323: Precompile: 0x06 ecAdd
Probably some of the wrong curve circuit logic would be useful here. https://github.com/privacy-scaling-explorations/halo2wrong

### PR DESCRIPTION
**Pull request by KimiWu123**
closed #323

### CODE REVIEW

**Comment by roynalnaruto** [comment]
@KimiWu123 Hi, is this PR ready for review?

File: src/zkevm_specs/evm_circuit/table.py [hunk 1]
```diff
@@ -548,6 +558,23 @@ class SigTableRow(TableRow):
     is_valid: FQ
 
 
+@dataclass(frozen=True)
+class EccTableRow(TableRow):
+    op_type: FQ
+    px: Word
+    py: Word
+    # qx is the scalar and qy must be zero if op_type is multiple
```
Comment (han0110) [inline src/zkevm_specs/evm_circuit/table.py:566]:
```suggestion
    # qx is the scalar and qy must be zero if op_type is Mul
```

File: src/zkevm_specs/evm_circuit/execution/precompiles/ecadd.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    # if is_success is false, outx and outy are zero
+    if is_success == FQ.zero():
+        instruction.constrain_zero(outx)
+        instruction.constrain_zero(outy)
+
+    # ecc table lookup
+    instruction.ecc_lookup(FQ(EccOpTag.Add), px, py, qx, qy, FQ.zero(), outx, outy, is_success)
+
+    # consume gas
```
Comment (han0110) [inline src/zkevm_specs/evm_circuit/execution/precompiles/ecadd.py:38]: duplicated?

File: src/zkevm_specs/ecc_circuit.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        precheck_p0y = cls.check_fq(p0[1].int_value())
+        precheck_p1x = cls.check_fq(p1[0].int_value())
+        precheck_p1y = cls.check_fq(p1[1].int_value())
+
+        point0 = (FP(p0[0].int_value()), FP(p0[1].int_value()))
+        point1 = (FP(p1[0].int_value()), FP(p1[1].int_value()))
+        is_valid_p0 = is_on_curve(point0, b)
+        is_valid_p1 = is_on_curve(point1, b)
+        is_infinite_p0 = is_inf(point0)
+        is_infinite_p1 = is_inf(point1)
```
Comment (han0110) [inline src/zkevm_specs/ecc_circuit.py:57]: It seems `is_on_curve` already checks if `is_inf`, but the `is_inf` returns `True` only if the input is `None`, so I think here they would never be `True`.

File: src/zkevm_specs/ecc_circuit.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    def verify_pairing(self, cs: ConstraintSystem):
+        # p and q are all zero. All input points are RLCed and stored in input_rlc
+        cs.constrain_zero(self.row.px)
+        cs.constrain_zero(self.row.py)
+        cs.constrain_zero(self.row.qx)
+        cs.constrain_zero(self.row.qy)
+        # output of pairing is either 0 or 1 and it stores in low part
+        cs.constrain_zero(self.row.out_x)
+        cs.constrain_bool(self.row.out_y)
+        cs.constrain_equal(self.row.out_y, self.row.is_valid)
```
Comment (han0110) [inline src/zkevm_specs/ecc_circuit.py:160]: I think pairing result and `is_valid` won't always be equal, but we could leave this to be fixed in future pairing PR.

File: tests/evm/precompiles/test_ecAdd.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    verify_steps,
+)
+from zkevm_specs.util import (
+    Word,
+    FQ,
+)
+from zkevm_specs.evm_circuit.table import EccOpTag, EccTableRow
+
+
+def gen_testing_data():
```
Comment (han0110) [inline tests/evm/precompiles/test_ecAdd.py:23]: It would be good to have testing data about point at infinity.

File: tests/test_ecc_circuit.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            out=(
+                0x030644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFD0,
+                0x15ED738C0E0A7C92E7845F96B2AE9C0A68A6A449E3538FC7FF3EBF7A5A18A2C4,
+            ),
+        ),
+        False,
+    )
+    # q = -p
+    # py_ecc doesn't support this case, it returns (0x30644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFD45, 1)
+    # p_plus_neg_p = (
```
Comment (han0110) [inline tests/test_ecc_circuit.py:65]: `py_ecc` should support this case, it represents point at infinity by `None`, see https://github.com/ethereum/py_ecc/blob/master/py_ecc/bn128/bn128_curve.py#L92

**Review by han0110 — COMMENTED** [review]
Overall LGTM!

File: src/zkevm_specs/ecc_circuit.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+        cs.constrain_equal(FQ(self.ecc_chip.verify_mul()), self.row.is_valid)
+
+    def verify_pairing(self, cs: ConstraintSystem):
+        # p and q are all zero. All input points are RLCed and stored in input_rlc
+        cs.constrain_zero(self.row.px)
+        cs.constrain_zero(self.row.py)
+        cs.constrain_zero(self.row.qx)
+        cs.constrain_zero(self.row.qy)
+        # output of pairing is either 0 or 1 and it stores in low part
```
Comment (miha-stopar) [inline src/zkevm_specs/ecc_circuit.py:157]: Nitpick: "and stored in the lower part"?

**Review by miha-stopar — COMMENTED** [review]
LGTM, but agree with Han's comments.

File: tests/evm/precompiles/test_ecAdd.py (hunk 5 above)
Comment (KimiWu123) [reply tests/evm/precompiles/test_ecAdd.py:23]: added in [9997f92](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/9997f929b03f2a1f8aa72fcfec98cc50611d13c0)

File: src/zkevm_specs/ecc_circuit.py (hunk 3 above)
Comment (KimiWu123) [reply src/zkevm_specs/ecc_circuit.py]: fixed in [b6c812d](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/b6c812d260d279ec95bbd8f9525d58dcbce75559)

File: src/zkevm_specs/evm_circuit/table.py [hunk 8]
```diff
@@ -350,6 +350,16 @@ def from_account_field_tag(field_tag: AccountFieldTag) -> MPTProofType:
...
 
 
+class EccOpTag(IntEnum):
+    """
+    Tag for EccTable that specifies the operation over ECC
+    """
+
+    Add = auto()  # addition of two EC points
+    Mul = auto()  # multiplication of two EC points
+    Pairing = auto()  # pairing of two EC points
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/table.py]: fixed by Rohit's comment

File: src/zkevm_specs/ecc_circuit.py (hunk 7 above)
Comment (KimiWu123) [reply src/zkevm_specs/ecc_circuit.py]: fixed in [bece499](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/bece499b09db05ca3ce5b2339a24a6ca37690427)

File: specs/precompile/06ecAdd.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+Two points are recovered from input. The field is expressed as `32` bytes for each and the input includes two points so the input size is `128` bytes.
+
+```
+input[0; 31] (32 bytes): x1
+input[32; 63] (32 bytes): y1
+input[64; 95] (32 bytes): x2
+input[96; 128] (32 bytes): y2
+```
+
+These two points are added and the result is returned. The result size is `64` bytes and $x$ and $y$ are montgomery form.
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply specs/precompile/06ecAdd.md]: fixed in [bece499](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/bece499b09db05ca3ce5b2339a24a6ca37690427)

File: tests/test_ecc_circuit.py (hunk 6 above)
Comment (KimiWu123) [reply tests/test_ecc_circuit.py]: fixed in [c1d5840](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/c1d5840498ec3fe3bc0a29d516f0944399dadcb9)

File: src/zkevm_specs/evm_circuit/execution/precompiles/ecadd.py (hunk 2 above)
Comment (KimiWu123) [reply src/zkevm_specs/evm_circuit/execution/precompiles/ecadd.py]: remvoed in [b6c812d](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/b6c812d260d279ec95bbd8f9525d58dcbce75559)

File: src/zkevm_specs/ecc_circuit.py (hunk 4 above)
Comment (KimiWu123) [reply src/zkevm_specs/ecc_circuit.py:171]: fixed in [b6c812d](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/500/commits/b6c812d260d279ec95bbd8f9525d58dcbce75559)

**Comment by KimiWu123** [comment]
> @KimiWu123 Hi, is this PR ready for review?

Sorry for the late reply, I was on vocation two weeks ago. It's ready for review now!

**Review by han0110 — APPROVED** [review]
All LGTM now! Nice work!

**Review by miha-stopar — APPROVED** [review]
LGTM!

**Pull request merged**